core/src/provider.rs

```diff
@@ -679,18 +635,43 @@ pub async fn create_client(
         (rpc_client, provider)
     };
 
-    Ok(Arc::new(
-        JsonRpcCachedProvider::new(
-            provider,
+    let real_rpc_chain_id = provider.get_chain_id().await.map_err(|e| {
+        RetryClientError::CouldNotConnectClient(RpcError::LocalUsageError(Box::new(e)))
+    })?;
+
+    if real_rpc_chain_id != chain_id {
+        return Err(RetryClientError::InvalidClientChainId(
```

Comment:
adding this check to make sure config `chain_id` matches the real RPC `chain_id`.
Tested this locally and if there is a mismatch indexer is going into a weird broken state.